### PR TITLE
CP-18172: Add license checks for nested-virt

### PR DESCRIPTION
**Pull request by jonludlam**
License is only checked on VM start.

### CODE REVIEW

File: ocaml/xapi/xapi_vm_helpers.ml [hunk 1]
```diff
@@ -41,6 +41,45 @@ let compute_memory_overhead ~__context ~vm =
...
+let check_restricted_platform_flags ~__context platform =
+  (* Takes a key and a predicate (string -> bool). If the key has an
+     associated value in the platform map, and the predicate is true
+     for that value, assert that the feature is enabled. *)
+  let check_key key predicate f =
+    if List.mem_assoc key platform && predicate (List.assoc key platform)
+    then Pool_features.assert_enabled ~__context ~f
+  in
+
+  let enabled_in_xenguest v =
```
Comment (lindig) [inline ocaml/xapi/xapi_vm_helpers.ml]: I think this code should be accessible from the outside to make sure we can use it in the VM lifecycle checks where currently there is a different implementation.

**Comment by lindig** [comment]
This looks good. I think the logic for checking `platform:nested-virt:value` should be shared.

File: ocaml/xapi/xapi_vm_helpers.ml (hunk 1 above)
Comment (jonludlam) [inline ocaml/xapi/xapi_vm_helpers.ml]: Agreed.

File: ocaml/xapi/xapi_vm.ml [hunk 2]
```diff
@@ -195,6 +195,9 @@ let start ~__context ~vm ~start_paused ~force =
 		debug "Setting ha_always_run on vm=%s as true during VM.start" (Ref.string_of vm)
 	end;
 
+	(* Check to see if we're using any restricted platform kvs. This raises
+	   an exception if so *)
+	Xapi_vm_helpers.check_restricted_platform_flags ~__context vmr.API.vM_platform;
```
Comment (robhoes) [inline ocaml/xapi/xapi_vm.ml]: Isn't this better done in Xapi_xenopsd when parsing the platform map? You would also handle the VM.resume case that way. And there is already some code there that manipulates the platform map.

Comment (jonludlam) [inline ocaml/xapi/xapi_vm.ml]: We explicitly don't want to do the check on VM.resume, just VM start, as you can always flip the switch and then start your VM without loss of data (just functionality) whereas the VM would have to be hard_shutdown to make it work if we did it on resume.

Having said that, I think it's better to keep the related functions together, so I think you're right about the location of the code. I'll move it.

**Comment by jonludlam** [comment]
Note, the last commit changes the semantics of `Xapi_xenops.Platform.is_true` - if the value associated with the key cannot be parsed, it will now raise an exception rather than use the default. The default is only used if the key is not present in the platform map.

File: ocaml/xapi/features.mli [hunk 3]
```diff
@@ -50,6 +50,7 @@ type feature =
 	| Guest_ip_setting             (** Enable use of Guest ip seting *)
 	| AD                           (** Enable use of Active Directory *)
 	| Ssl_legacy_switch            (** Enable the control switch for SSL/TLS legacy-mode. *)
+        | Nested_virt                  (** Enable the use of nested virtualisation *)
```
Comment (lindig) [inline ocaml/xapi/features.mli]: indentation is off

**Comment by lindig** [comment]
I've flagged a minor inconsistency in the indentation - otherwise I'd say we merge this.

**Comment by robhoes** [comment]
Ok, I'm happy with it after squashing patches 4 and 5 into 2, and fixing the indentation.

**Comment by jonludlam** [comment]
I've been mulling over the change in semantics, and have an alternative proposal. I've also squashed the commits. Will now force push over this PR.

**Comment by robhoes** [comment]
:shipit:

**Comment by lindig** [comment]
:shipit:

**Comment by jonludlam** [comment]
@xen-git retest this please

**Comment by jonludlam** [comment]
@xen-git retest this please

**Comment by robhoes** [comment]
I am going to revert this, because the patch broke VM resume and migrate, and turned out not quite right anyway.

**Pull request merged**